File: 10-simulation/static_sim/play_requests.py
import sys  
import time  
import os
import requests  
from concurrent.futures import ThreadPoolExecutor  
  
def send_request(port):  
    start_time = time.time()  
    requests.post(f"http://127.0.0.1:{port}/fibonacci", json={"number": 26})  
    end_time = time.time()  
    response_time = end_time - start_time  
    return response_time  
  
def start_play(filename, port):  
    line_number = 0  
    filepath = f'''{os.path.dirname(__file__)}/{filename}'''
    with open(filepath, 'r') as file:  
        for line in file:  
            line_number += 1  
            concurrent_requests = int(float(line.strip()))  
            if concurrent_requests > 0:  
                # Send concurrent_requests requests in parallel  
                with ThreadPoolExecutor(max_workers=concurrent_requests) as executor:  
                    response_times = list(executor.map(send_request, [port]*concurrent_requests))  
                average_response_time = sum(response_times) / len(response_times)  
                print(f"Line {line_number}: {concurrent_requests} requests take average response time: {average_response_time} seconds")  
            # Sleep before starting the next round of requests  
            time.sleep(1)  
  
if __name__ == "__main__":  
    if len(sys.argv) != 3:  
        print(f"Usage: {sys.argv[0]} <filename> <port>")  
        sys.exit(1)  
  
    start_play(sys.argv[1], sys.argv[2])  

File: 10-simulation/static_sim/play_requests.go
package main  
  
import (  
	"bufio"  
	"bytes"
	"fmt"  
	"net/http"  
	"os"  
	"strconv"  
	"sync"  
	"time"  
)  
  
func sendRequest(port string, wg *sync.WaitGroup) {  
	defer wg.Done()  
	// startTime := time.Now()  
	jsonStr := []byte(`{"number":30}`)  
	http.Post("http://127.0.0.1:"+port+"/fibonacci", "application/json", bytes.NewBuffer(jsonStr))  
	// endTime := time.Now()  
	// responseTime := endTime.Sub(startTime)  
	// fmt.Println("Response time: ", responseTime.Seconds(), " seconds")  
}  
  
func startPlay(filename, port string) {  
	file, err := os.Open(filename)  
	if err != nil {  
		fmt.Println(err)  
		return  
	}  
	fmt.Println(file)
	defer file.Close()  
  
	scanner := bufio.NewScanner(file)  
	lineNumber := 0  
	for scanner.Scan() {  
		lineNumber++  
		floatVal, err := strconv.ParseFloat(scanner.Text(), 64)  
		if err != nil {  
			fmt.Println("Failed to parse float: ", err)  
			return  
		}  
		concurrentRequests := int(floatVal)  

		if concurrentRequests > 0 {  
			fmt.Println("Line number: ", lineNumber, " Value: ", concurrentRequests)
			var wg sync.WaitGroup  
			for i := 0; i < concurrentRequests; i++ {  
				wg.Add(1)  
				go sendRequest(port, &wg)  
			}  
			wg.Wait()  
		}  
		time.Sleep(1 * time.Second)  
	}  
}  
  
func main() {  
	if len(os.Args) != 3 {  
		fmt.Println("Usage: ", os.Args[0], " <filename> <port>")  
		os.Exit(1)  
	}  
	startPlay(os.Args[1], os.Args[2])  
}  
